File: targets/targets.go
// Package targets provides types for representing archery targets.
package targets

import (
	"fmt"
	"math"
	"sort"
	"strings"

	"github.com/retbrown/archeryutils/length"
)

// ScoringSystem names a target face/scoring system.
type ScoringSystem string

// All scoring systems supported natively.
const (
	FiveZone              ScoringSystem = "5_zone"
	TenZone               ScoringSystem = "10_zone"
	TenZoneCompound       ScoringSystem = "10_zone_compound"
	TenZoneSixRing        ScoringSystem = "10_zone_6_ring"
	TenZoneFiveRing       ScoringSystem = "10_zone_5_ring"
	TenZoneFiveRingCmpd   ScoringSystem = "10_zone_5_ring_compound"
	ElevenZone            ScoringSystem = "11_zone"
	ElevenZoneSixRing     ScoringSystem = "11_zone_6_ring"
	ElevenZoneFiveRing    ScoringSystem = "11_zone_5_ring"
	WAField               ScoringSystem = "WA_field"
	IFAAField             ScoringSystem = "IFAA_field"
	IFAAFieldExpert       ScoringSystem = "IFAA_field_expert"
	AANationalField       ScoringSystem = "AA_national_field"
	BeiterHitMiss         ScoringSystem = "Beiter_hit_miss"
	Worcester             ScoringSystem = "Worcester"
	Worcester2Ring        ScoringSystem = "Worcester_2_ring"
	Custom                ScoringSystem = "Custom"
)

// supportedSystems lists every scoring system archeryutils knows by default.
var supportedSystems = []ScoringSystem{
	FiveZone, TenZone, TenZoneCompound, TenZoneSixRing, TenZoneFiveRing,
	TenZoneFiveRingCmpd, ElevenZone, ElevenZoneSixRing, ElevenZoneFiveRing,
	WAField, IFAAField, IFAAFieldExpert, AANationalField, BeiterHitMiss,
	Worcester, Worcester2Ring, Custom,
}

var supportedSystemSet map[ScoringSystem]struct{}

func init() {
	supportedSystemSet = make(map[ScoringSystem]struct{}, len(supportedSystems))
	for _, s := range supportedSystems {
		supportedSystemSet[s] = struct{}{}
	}
}

// FaceSpec maps each ring diameter (in metres) to its score value.
type FaceSpec map[float64]int

// Quantity is a scalar value with its unit.
type Quantity struct {
	Value float64
	Units string
}

// CM returns a Quantity in centimetres.
func CM(v float64) Quantity { return Quantity{v, "cm"} }

// Metres returns a Quantity in metres.
func Metres(v float64) Quantity { return Quantity{v, "metre"} }

// Yards returns a Quantity in yards.
func Yards(v float64) Quantity { return Quantity{v, "yard"} }

// Inches returns a Quantity in inches.
func Inches(v float64) Quantity { return Quantity{v, "inch"} }

// supported unit sets for diameter and distance
var (
	supportedDiamUnits = length.Union(length.CM, length.Inch, length.Metre)
	supportedDistUnits = length.Union(length.Yard, length.Metre)
)

// Target represents an archery target face at a given distance.
type Target struct {
	diameter       float64  // metres
	distance       float64  // metres
	nativeDiameter Quantity // original value+unit
	nativeDistance Quantity // original value+unit
	scoringSystem  ScoringSystem
	indoor         bool
	faceSpec       FaceSpec
}

// NewTarget creates a Target. diameter defaults to centimetres; distance defaults to metres.
func NewTarget(system ScoringSystem, diameter, distance Quantity, indoor bool) (*Target, error) {
	if _, ok := supportedSystemSet[system]; !ok {
		names := make([]string, len(supportedSystems))
		for i, s := range supportedSystems {
			names[i] = "'" + string(s) + "'"
		}
		return nil, fmt.Errorf(
			"Invalid Target Face Type specified.\nPlease select from %s.",
			strings.Join(names, ", "),
		)
	}

	diamVal, diamUnit, err := length.ParseOptionalUnitsFloat(
		diameter.Value, supportedDiamUnits, diameter.Units,
	)
	if err != nil {
		return nil, err
	}
	distVal, distUnit, err := length.ParseOptionalUnitsFloat(
		distance.Value, supportedDistUnits, distance.Units,
	)
	if err != nil {
		return nil, err
	}

	diamM, err := length.ToMetres(diamVal, diamUnit)
	if err != nil {
		return nil, err
	}
	distM, err := length.ToMetres(distVal, distUnit)
	if err != nil {
		return nil, err
	}

	t := &Target{
		scoringSystem:  system,
		diameter:       diamM,
		nativeDiameter: Quantity{diamVal, diamUnit},
		distance:       distM,
		nativeDistance: Quantity{distVal, distUnit},
		indoor:         indoor,
	}

	if system != Custom {
		spec, err := GenFaceSpec(system, diamM)
		if err != nil {
			return nil, err
		}
		t.faceSpec = spec
	}

	return t, nil
}

// FromFaceSpec creates a Target with a custom scoring spec.
// spec may be a FaceSpec (units default to metres) or [2]any{FaceSpec, string}.
func FromFaceSpec(spec any, diameter, distance Quantity, indoor bool) (*Target, error) {
	var rawSpec FaceSpec
	var specUnits string

	switch v := spec.(type) {
	case FaceSpec:
		rawSpec = v
		specUnits = "metre"
	case [2]any:
		fs, ok := v[0].(FaceSpec)
		if !ok {
			return nil, fmt.Errorf("first element must be FaceSpec, got %T", v[0])
		}
		u, ok := v[1].(string)
		if !ok {
			return nil, fmt.Errorf("second element must be string unit, got %T", v[1])
		}
		rawSpec = fs
		specUnits = u
	default:
		return nil, fmt.Errorf("spec must be FaceSpec or [2]any{FaceSpec,string}, got %T", spec)
	}

	// Validate spec units
	defSpecUnit, err := validateAndResolveUnit(specUnits, supportedDiamUnits)
	if err != nil {
		return nil, err
	}

	// Convert spec ring diameters to metres
	converted := make(FaceSpec, len(rawSpec))
	for ringDiam, score := range rawSpec {
		m, err := length.ToMetres(ringDiam, defSpecUnit)
		if err != nil {
			return nil, err
		}
		converted[rnd6(m)] = score
	}

	t, err := NewTarget(Custom, diameter, distance, indoor)
	if err != nil {
		return nil, err
	}
	t.faceSpec = converted
	return t, nil
}

func validateAndResolveUnit(unit string, supported length.UnitSet) (string, error) {
	if _, ok := supported[unit]; !ok {
		return "", fmt.Errorf("Unit %q not recognised.", unit)
	}
	def, _ := length.DefinitiveUnit(unit)
	return def, nil
}

// ScoringSystem returns the scoring system of this target.
func (t *Target) ScoringSystem() ScoringSystem { return t.scoringSystem }

// Diameter returns the target diameter in metres.
func (t *Target) Diameter() float64 { return t.diameter }

// Distance returns the target distance in metres.
func (t *Target) Distance() float64 { return t.distance }

// NativeDiameter returns diameter in its original unit.
func (t *Target) NativeDiameter() Quantity { return t.nativeDiameter }

// NativeDistance returns distance in its original unit.
func (t *Target) NativeDistance() Quantity { return t.nativeDistance }

// Indoor reports whether the target is marked as indoor.
func (t *Target) Indoor() bool { return t.indoor }

// FaceSpec returns a copy of the target's face spec.
// Returns an error if the target was created as Custom without a valid spec.
func (t *Target) FaceSpec() (FaceSpec, error) {
	if t.faceSpec == nil {
		return nil, fmt.Errorf(
			"Trying to generate face spec for custom target " +
				"but no existing spec found: " +
				"try instantiating with FromFaceSpec instead",
		)
	}
	out := make(FaceSpec, len(t.faceSpec))
	for k, v := range t.faceSpec {
		out[k] = v
	}
	return out, nil
}

// MaxScore returns the maximum score achievable on this target face.
func (t *Target) MaxScore() float64 {
	var max float64
	for _, v := range t.faceSpec {
		if float64(v) > max {
			max = float64(v)
		}
	}
	return max
}

// MinScore returns the minimum score achievable on this target face (excluding miss).
func (t *Target) MinScore() float64 {
	if len(t.faceSpec) == 0 {
		return 0
	}
	min := math.MaxFloat64
	for _, v := range t.faceSpec {
		if float64(v) < min {
			min = float64(v)
		}
	}
	return min
}

// String returns a human-readable representation of the Target.
func (t *Target) String() string {
	d, du := t.nativeDiameter.Value, t.nativeDiameter.Units
	di, diu := t.nativeDistance.Value, t.nativeDistance.Units
	return fmt.Sprintf("Target('%s', (%.6g, '%s'), (%.6g, '%s'), indoor=%v)",
		t.scoringSystem, d, du, di, diu, t.indoor)
}

// Equal reports whether two Targets are equivalent.
func (t *Target) Equal(other *Target) bool {
	if t == other {
		return true
	}
	if other == nil {
		return false
	}
	if t.scoringSystem != other.scoringSystem ||
		t.diameter != other.diameter ||
		t.distance != other.distance ||
		t.nativeDiameter != other.nativeDiameter ||
		t.nativeDistance != other.nativeDistance ||
		t.indoor != other.indoor {
		return false
	}
	// Compare face specs
	if len(t.faceSpec) != len(other.faceSpec) {
		return false
	}
	for k, v := range t.faceSpec {
		if ov, ok := other.faceSpec[k]; !ok || ov != v {
			return false
		}
	}
	return true
}

// GenFaceSpec derives the face specification for supported scoring systems.
// diameter must be in metres.
func GenFaceSpec(system ScoringSystem, diameter float64) (FaceSpec, error) {
	removedRings := map[ScoringSystem]int{
		TenZoneSixRing:      4,
		TenZoneFiveRing:     5,
		TenZoneFiveRingCmpd: 5,
		ElevenZoneSixRing:   4,
		ElevenZoneFiveRing:  5,
		Worcester2Ring:      3,
	}

	missing := removedRings[system]
	spec := make(FaceSpec)

	switch system {
	case FiveZone:
		for n := 1; n <= 10; n += 2 {
			spec[rnd6(float64(n+1)*diameter/10)] = 10 - n
		}

	case TenZone, TenZoneSixRing, TenZoneFiveRing:
		for n := 1; n <= 10-missing; n++ {
			spec[rnd6(float64(n)*diameter/10)] = 11 - n
		}

	case TenZoneCompound, TenZoneFiveRingCmpd:
		spec[rnd6(diameter/20)] = 10
		for n := 2; n <= 10-missing; n++ {
			spec[rnd6(float64(n)*diameter/10)] = 11 - n
		}

	case ElevenZone, ElevenZoneSixRing, ElevenZoneFiveRing:
		spec[rnd6(diameter/20)] = 11
		for n := 1; n <= 10-missing; n++ {
			spec[rnd6(float64(n)*diameter/10)] = 11 - n
		}

	case WAField:
		spec[rnd6(diameter/10)] = 6
		for n := 1; n <= 5; n++ {
			spec[rnd6(float64(n)*diameter/5)] = 6 - n
		}

	case IFAAField:
		for _, n := range []int{1, 3, 5} {
			spec[rnd6(float64(n)*diameter/5)] = 5 - n/2
		}

	case AANationalField:
		for n := 1; n <= 5; n++ {
			spec[rnd6(float64(n)*diameter/5)] = 6 - n
		}

	case BeiterHitMiss:
		spec[diameter] = 1

	case Worcester, Worcester2Ring, IFAAFieldExpert:
		for n := 1; n <= 5-missing; n++ {
			spec[rnd6(float64(n)*diameter/5)] = 6 - n
		}

	default:
		return nil, fmt.Errorf("Scoring system %q is not supported", system)
	}

	return spec, nil
}

// rnd6 rounds a float64 to 6 decimal places, matching Python's round(x, ndigits=6).
func rnd6(x float64) float64 {
	return math.Round(x*1e6) / 1e6
}

// SortedKeys returns the keys of a FaceSpec in ascending order (for deterministic iteration).
func SortedKeys(fs FaceSpec) []float64 {
	keys := make([]float64, 0, len(fs))
	for k := range fs {
		keys = append(keys, k)
	}
	sort.Float64s(keys)
	return keys
}

File: ts/src/targets.ts
import { toMetres, definitiveUnit, SUPPORTED_DIAM_UNITS, SUPPORTED_DIST_UNITS } from './length.js';

export type ScoringSystem =
  | '5_zone' | '10_zone' | '10_zone_compound' | '10_zone_6_ring'
  | '10_zone_5_ring' | '10_zone_5_ring_compound' | '11_zone'
  | '11_zone_6_ring' | '11_zone_5_ring' | 'WA_field' | 'IFAA_field'
  | 'IFAA_field_expert' | 'AA_national_field' | 'Beiter_hit_miss'
  | 'Worcester' | 'Worcester_2_ring' | 'Custom';

export const ScoringSystem = {
  FiveZone: '5_zone' as ScoringSystem,
  TenZone: '10_zone' as ScoringSystem,
  TenZoneCompound: '10_zone_compound' as ScoringSystem,
  TenZoneSixRing: '10_zone_6_ring' as ScoringSystem,
  TenZoneFiveRing: '10_zone_5_ring' as ScoringSystem,
  TenZoneFiveRingCmpd: '10_zone_5_ring_compound' as ScoringSystem,
  ElevenZone: '11_zone' as ScoringSystem,
  ElevenZoneSixRing: '11_zone_6_ring' as ScoringSystem,
  ElevenZoneFiveRing: '11_zone_5_ring' as ScoringSystem,
  WAField: 'WA_field' as ScoringSystem,
  IFAAField: 'IFAA_field' as ScoringSystem,
  IFAAFieldExpert: 'IFAA_field_expert' as ScoringSystem,
  AANationalField: 'AA_national_field' as ScoringSystem,
  BeiterHitMiss: 'Beiter_hit_miss' as ScoringSystem,
  Worcester: 'Worcester' as ScoringSystem,
  Worcester2Ring: 'Worcester_2_ring' as ScoringSystem,
  Custom: 'Custom' as ScoringSystem,
} as const;

const SUPPORTED_SYSTEMS = new Set<string>([
  '5_zone', '10_zone', '10_zone_compound', '10_zone_6_ring',
  '10_zone_5_ring', '10_zone_5_ring_compound', '11_zone',
  '11_zone_6_ring', '11_zone_5_ring', 'WA_field', 'IFAA_field',
  'IFAA_field_expert', 'AA_national_field', 'Beiter_hit_miss',
  'Worcester', 'Worcester_2_ring', 'Custom',
]);

/** Maps each ring diameter (metres) to its score value. */
export type FaceSpec = Map<number, number>;

export interface Quantity {
  value: number;
  units: string;
}

export function cm(v: number): Quantity { return { value: v, units: 'cm' }; }
export function metres(v: number): Quantity { return { value: v, units: 'metre' }; }
export function yards(v: number): Quantity { return { value: v, units: 'yard' }; }
export function inches(v: number): Quantity { return { value: v, units: 'inch' }; }

function rnd6(x: number): number {
  return Math.round(x * 1e6) / 1e6;
}

function resolveQuantity(q: Quantity, supportedAliases: Set<string>): [number, string] {
  const unitAlias = q.units || 'cm';
  if (!supportedAliases.has(unitAlias)) {
    throw new Error(`Unit "${unitAlias}" not recognised.`);
  }
  const def = definitiveUnit(unitAlias);
  return [q.value, def];
}

export class Target {
  private constructor(
    private readonly _diameter: number,       // metres
    private readonly _distance: number,       // metres
    private readonly _nativeDiameter: Quantity,
    private readonly _nativeDistance: Quantity,
    private readonly _scoringSystem: ScoringSystem,
    private readonly _indoor: boolean,
    private readonly _faceSpec: FaceSpec,
  ) {}

  get scoringSystem(): ScoringSystem { return this._scoringSystem; }
  get diameter(): number { return this._diameter; }
  get distance(): number { return this._distance; }
  get nativeDiameter(): Quantity { return this._nativeDiameter; }
  get nativeDistance(): Quantity { return this._nativeDistance; }
  get indoor(): boolean { return this._indoor; }

  get faceSpec(): FaceSpec {
    if (!this._faceSpec) {
      throw new Error(
        'No face spec available for custom target — use Target.fromFaceSpec() instead'
      );
    }
    return new Map(this._faceSpec);
  }

  get maxScore(): number {
    let max = 0;
    for (const v of this._faceSpec.values()) {
      if (v > max) max = v;
    }
    return max;
  }

  get minScore(): number {
    let min = Infinity;
    for (const v of this._faceSpec.values()) {
      if (v < min) min = v;
    }
    return min === Infinity ? 0 : min;
  }

  static create(
    system: ScoringSystem,
    diameter: Quantity,
    distance: Quantity,
    indoor: boolean,
  ): Target {
    if (!SUPPORTED_SYSTEMS.has(system)) {
      throw new Error(
        `Invalid Target Face Type specified. Please select from ${[...SUPPORTED_SYSTEMS].map(s => `'${s}'`).join(', ')}.`
      );
    }
    const [diamVal, diamUnit] = resolveQuantity(diameter, SUPPORTED_DIAM_UNITS);
    const [distVal, distUnit] = resolveQuantity(distance, SUPPORTED_DIST_UNITS);
    const diamM = toMetres(diamVal, diamUnit);
    const distM = toMetres(distVal, distUnit);
    const spec = system !== 'Custom' ? genFaceSpec(system, diamM) : new Map<number, number>();
    return new Target(
      diamM, distM,
      { value: diamVal, units: diamUnit },
      { value: distVal, units: distUnit },
      system, indoor, spec,
    );
  }

  static fromFaceSpec(
    spec: FaceSpec | [FaceSpec, string],
    diameter: Quantity,
    distance: Quantity,
    indoor: boolean,
  ): Target {
    let rawSpec: FaceSpec;
    let specUnit: string;
    if (spec instanceof Map) {
      rawSpec = spec;
      specUnit = 'metre';
    } else {
      [rawSpec, specUnit] = spec;
    }
    const def = definitiveUnit(specUnit);
    const converted = new Map<number, number>();
    for (const [ringDiam, score] of rawSpec) {
      converted.set(rnd6(toMetres(ringDiam, def)), score);
    }
    const base = Target.create('Custom', diameter, distance, indoor);
    return new Target(
      base._diameter, base._distance,
      base._nativeDiameter, base._nativeDistance,
      'Custom', indoor, converted,
    );
  }

  equals(other: Target): boolean {
    if (this === other) return true;
    if (
      this._scoringSystem !== other._scoringSystem ||
      this._diameter !== other._diameter ||
      this._distance !== other._distance ||
      this._nativeDiameter.value !== other._nativeDiameter.value ||
      this._nativeDiameter.units !== other._nativeDiameter.units ||
      this._nativeDistance.value !== other._nativeDistance.value ||
      this._nativeDistance.units !== other._nativeDistance.units ||
      this._indoor !== other._indoor ||
      this._faceSpec.size !== other._faceSpec.size
    ) return false;
    for (const [k, v] of this._faceSpec) {
      if (other._faceSpec.get(k) !== v) return false;
    }
    return true;
  }

  toString(): string {
    const { value: d, units: du } = this._nativeDiameter;
    const { value: di, units: diu } = this._nativeDistance;
    return `Target('${this._scoringSystem}', (${d}, '${du}'), (${di}, '${diu}'), indoor=${this._indoor})`;
  }
}

export function genFaceSpec(system: ScoringSystem, diameter: number): FaceSpec {
  const spec = new Map<number, number>();
  const REMOVED: Partial<Record<ScoringSystem, number>> = {
    '10_zone_6_ring': 4,
    '10_zone_5_ring': 5,
    '10_zone_5_ring_compound': 5,
    '11_zone_6_ring': 4,
    '11_zone_5_ring': 5,
    'Worcester_2_ring': 3,
  };
  const missing = REMOVED[system] ?? 0;

  switch (system) {
    case '5_zone':
      for (let n = 1; n <= 10; n += 2) {
        spec.set(rnd6((n + 1) * diameter / 10), 10 - n);
      }
      break;
    case '10_zone':
    case '10_zone_6_ring':
    case '10_zone_5_ring':
      for (let n = 1; n <= 10 - missing; n++) {
        spec.set(rnd6(n * diameter / 10), 11 - n);
      }
      break;
    case '10_zone_compound':
    case '10_zone_5_ring_compound':
      spec.set(rnd6(diameter / 20), 10);
      for (let n = 2; n <= 10 - missing; n++) {
        spec.set(rnd6(n * diameter / 10), 11 - n);
      }
      break;
    case '11_zone':
    case '11_zone_6_ring':
    case '11_zone_5_ring':
      spec.set(rnd6(diameter / 20), 11);
      for (let n = 1; n <= 10 - missing; n++) {
        spec.set(rnd6(n * diameter / 10), 11 - n);
      }
      break;
    case 'WA_field':
      spec.set(rnd6(diameter / 10), 6);
      for (let n = 1; n <= 5; n++) {
        spec.set(rnd6(n * diameter / 5), 6 - n);
      }
      break;
    case 'IFAA_field':
      for (const n of [1, 3, 5]) {
        spec.set(rnd6(n * diameter / 5), 5 - Math.floor(n / 2));
      }
      break;
    case 'AA_national_field':
      for (let n = 1; n <= 5; n++) {
        spec.set(rnd6(n * diameter / 5), 6 - n);
      }
      break;
    case 'Beiter_hit_miss':
      spec.set(diameter, 1);
      break;
    case 'Worcester':
    case 'Worcester_2_ring':
    case 'IFAA_field_expert':
      for (let n = 1; n <= 5 - missing; n++) {
        spec.set(rnd6(n * diameter / 5), 6 - n);
      }
      break;
    default:
      throw new Error(`Scoring system "${system}" is not supported`);
  }
  return spec;
}

export function sortedKeys(fs: FaceSpec): number[] {
  return [...fs.keys()].sort((a, b) => a - b);
}
